### ISSUE TITLE
System.LimitException: dlrs:Too many DML rows: 10001

### ISSUE THREAD

**Issue by venkatasatishp** [afawcett/declarative-lookup-rollup-summaries#731]
Hi,
We are doing data load into a custom object and all our rollups are scheduled ,none of them are realtime.
we are getting the following error when doing dataload (Bulk load) :CANNOT_INSERT_UPDATE_ACTIVATE_ENTITY
OBJ: Custom_Object__c - dlrs_Custom_ObjectTrigger: System.LimitException: dlrs:Too many DML rows: 10001 .

can you @afawcett please advise how to resolve this issue.

**Comment by aheber** [comment]
@venkatasatishp. I'm going to make some assumptions and you can tell me where I hit or miss.

That error comes from trying to update/insert too many rows into the database during one single transaction. Salesforce only lets you process up to 10K rows per-transaction. Normally you're dealing with batch sizes of 200 records during your bulk load but that can be multiplied.

If DLRS is solely the cause that means you would have to have 51+ active scheduled jobs all trying to push records into Salesforce for DLRS to consume and pointers to records that should be processed later.

In short there is nothing you can do if you are configured like that. You either have to disable some jobs and run full scheduled processes later, or you might want to consider setting some rollups to real-time. Depending on your situation that might perform better because it doesn't need to leave behind markers for all the records that _might_ need to be processed later.

**Comment by venkatasatishp** [comment]
@aheber  My Issue is not with Scheduled Job . 

My issue is happening when i am doing data load (Bulk load) to child object and none of my rollup are real-time . 
The trigger need to come out of the loop if none of the rollup are real-time. 
@afawcett I am hoping the trigger got a query to verify  how many scheduled or real-time . If No "Real-time" Job just come out of the trigger .   I see the load is failing on inserts not on Updates. Yesterday night we did 200 child records loaded all 200 failed with same error.

**Comment by aheber** [comment]
@venkatasatishp I don't think you're understanding what I'm explaining.

A DLRS job running in scheduled mode still utilizes the trigger and monitors the records for meaningful changes. When it sees a change it adds a record to the database for each DLRS job that this particular record qualified to fire a recalculation for, that needs to be processed later, this behaves as a queue for the scheduled job to know which records need to be processed when it runs. When you run in bulk it increases the number of records that need to be added to the "queue" and if it tries to add too many records (number of records * number of jobs) then it crashes all of the records in that update request, not just specific records. This is because the entire transaction tried manipulating more than 10K records (trying to insert queue records).

It is possible there is something else in your system contributing to this error but to my knowledge this is the only scenario where DLRS can cause this error. 

To manage this you'll need to review your rollups and either temporarily disable some and run full calculations later, or convert some to real-time as it will actually cause less direct DML. (Scheduled mode is generally useful when you have lots and lots of sibling records or when updating the parent would cause cascading DML and you need to isolate those transactions for governor/performance reasons).

If you would like DLRS to have zero impact on your inserts then you need to run full calculations on your jobs at some interval. Setting these to Active and Scheduled doesn't alleviate the trigger or defer all possible DML as the system still keeps track of what records to process later by way of DML.

**Comment by venkatasatishp** [comment]
@aheber  Thank you for Confirming.  Whom can i contact @DLRS to get permission to give Salesforce access and enable debug log on DLRS Managed package and Salesforce can review it for us.

**Comment by aheber** [comment]
As far as I know @afawcett is the only one who can speak on behalf of DLRS.

**Comment by venkatasatishp** [comment]
@afawcett   Salesforce support advised us to get permission  from the DLRS so they can check the debug logs related to the package .

**Comment by timmy2cakes** [comment]
Please can someone help me here I'm not getting any response on this #733

**Comment by afawcett** [comment]
@venkatasatishp i don't have anything more to add than @aheber here. If your code is doing 10k child record inserts, at minimum the overhead of an incremental schedule configured rollup will be some inserts into the tools own custom objects to queue up the rollups for later processing. All inserts in salesforce count towards 10k limit, your inserts and the tools. Your only option is to reduce the size of your batches.
